### DISCUSSION TITLE
Integrating a decision procedure written in ACL2s and lisp with Z3

### DISCUSSION THREAD

**Discussion by ankitku** [Z3Prover/z3#5495] [General]
We are working on integrating our string solver SeqSolve written in ACL2s and Lisp with Z3. SeqSolve solves both string and length constraints using another instance of Z3 as a backend (Linear Integer Arithmetic) LIA solver.

   The integration works by loading a modified version of Z3 containing a custom theory that we developed. This theory invokes SeqSolve and passes constraints as a SMTLIB string  using a function pointer. SeqSolve then uses Z3_eval_smtlib2_string to parse the SMT2LIB string it received and solves it. We have attached a diagram showing the SeqSolve stack below.

<img width="740" alt="Screenshot 2021-08-06 at 14 33 55" src="https://user-images.githubusercontent.com/519059/130143569-a1cb1e7e-b3bf-4c7f-a33e-143769aa36e2.png">

Our issue stems from the fact that we would like the C API Z3_solver that the Lisp function will end up using to have access to the constraints that exist in the original problem or have been added by other theories. One way we considered getting these constraints is by using the C API inside of the Lisp function. However, SeqSolve’s invocation of Z3_eval_smtlib2_string creates its own C++ solver instance, and it's not clear that it is possible to access all of the information needed to make a C Z3_solver_ref struct that refers to the "same" solver. We also tried using smt_context::get_assertions inside of our custom theory, but that simply dumps all the assertions seen in the input smt2 file. If there is any way to get a reference to the Z3_Solver from within a theory inside Z3, such that it can be passed on to the lisp function in an ffi call? The goal here is to copy the current Z3_solver being used inside a theory of Z3, and send the copy's reference for use by the lisp procedure so it can use it independently.

Thanks.

**Comment by mister-walter** [comment]
We also tried using `Z3_solver_to_string` and `Z3_solver_from_string`, but as far as I can remember, this did not include all of the information known by Z3 at the point where the custom theory was invoked.

**Comment by NikolajBjorner** [comment]
The solver interface is a general interface that can cover different backends with each their characteritics.
To force using just the SMT core solver you should initialize a Z3_mk_simple_solver. 
You can then run a query with the solver (say with a bounded number of conflicts by setting smt.max_conflicts).
To tease out the current set of base assertions and lemmas with low glue level you can _copy_ that solver into a fresh solver.
The function Z3_solver_translate supports copy routines between different or the same contexts. You can use the same context for your use.
Then the copied solver is initialized with assertions that are the result of solving. 
The original solver may only expose assertions that are added from outside.

You can for example try:

```
from z3 import *

set_option(verbose=10)
s = SimpleSolver()
x, y, z = Ints('x y z')
s.add(x == y, Or(x > y, x > z))
s.add(Or(x == y, x == z))
s.check()
s1 = s.translate(Context())
print(s1)
print(s)
```

You will then see some unusual state printed for both s and s1.
On the other hand, if you use just the standard "Solver" instead of "SimpleSolver", the internal state gets hidden and is then opaque.

**Reply by mister-walter** [reply]
Thanks for the reply!

Part of our issue is that we want to use Z3 with our custom tactic on SMTLIB2 files just like "vanilla" Z3, and so we are using `Z3_eval_smtlib2_string` on the contents of the relevant files after performing some setup. It does not seem as though it is possible to provide a `Solver` to this function for use when parsing the file, which makes sense as in the general case Z3 may need to make multiple `Solver`s depending on the commands inside of a SMTLIB2 file. We tried creating a custom `SolverFactory` that takes in a `Solver` and returns it every time it is asked for a new `Solver`, but that did not work properly.

We are currently using `Z3_solver_translate`, but as you noted the only assertions that are copied over are those that were added using an external interface (as opposed to those added by other theories etc). Is there no way to, given a `Solver a`, create a `Solver b` such that `b` contains all external assertions plus things like theory axioms and additional mid-solve information?

For example, assume that the file we're evaluating contains the assertion `(or P Q)`, where both `P` and `Q` are clauses that are relevant to our tactic. Assume that `P` is `UNSAT` and `Q` is `SAT`. Imagine then that Z3 "decides" to investigate the satisfiability of `P` first, and then `Q`. Other theories may assert some theory axioms based on the contents of `P`. Our theory may add some theory axioms and theory variables. Then, Z3 calls `new_eq_eh` for our theory. Let the `Solver` that is performing the current `check_sat` be `S` (assume that we're using a `SimpleSolver` as you suggest above). At this point, we would like to call out to our Lisp function, which ideally would use all of the information known to be true by `S` at this point.

The problem lies in the fact that `S` is currently in the middle of a `check_sat`, and it seems that Z3 does not allow us to directly use `S` with the C API when it is in the middle of a `check_sat`. (I'm eliding some details here regarding the difference between `solver` and `Z3_solver`, and how we might e.g. prevent infinite looping with our theory calling `check_sat` which calls our theory...) So it seems like we need to create a new `Solver`. This `Solver` should contain all information that Z3 knows about `P`, including any theory axioms we added (note that we need to add theory axioms since it's not legal to call `smt::context::assert_expr` inside a theory call). For our theory axioms we could get around this issue by keeping track of which theory axioms we add, and sending them over to be added to the `Z3_solver` on the Lisp side. However, I'm concerned about other information that Z3 may have (e.g. that `P` is not true).

Hopefully this makes sense. It's possible that Z3 is simply not designed to allow the kind of interaction we're looking for here, so apologies if that is the case.

**Comment by NikolajBjorner** [comment]
I am not sure what "new_eq_eh for our theory" means. 

The code I can find online https://github.com/ankitku/SeqSolve/tree/master/model-analyses/lisp-z3/src/z3 only references external APIs.
new_eq_eh is internal. 
The external API supports a specialized instance of user callbacks, but I don't see those mentioned in your lisp bindings.
The user callbacks are defined here: https://github.com/Z3Prover/z3/blob/aa0529895096d45238c1d2f25d89d883d30c1192/src/api/z3_api.h#L6541
They can be used to pick up variable assignments to Boolean and Bit-vectors and allow callbacks.
The Python bindings go through the hoops of registering callback functions. 
An application of this newer API is sketched here: http://theory.stanford.edu/~nikolaj/supercharging.html
It is generally suitable for supporting ad-hoc global propagators.

**Reply by mister-walter** [reply]
Apologies for any confusion!
The code in the `seqsolve` repo is old - we're working in https://github.com/ankitku/z3/tree/paper (note the `paper` branch)

We have a subclass of `smt::theory` that we have added (https://github.com/ankitku/z3/blob/paper/src/smt/theory_transeq.cpp) and Z3 calls its `new_eq_eh` method. Just to be clear, we have added two things here:
1. A `theory` subclass that uses the internal `theory` APIs to e.g. add theory axioms and variables. We want this to call out to the Lisp code, described below, and use the response of the Lisp code.
2. Some Lisp code that uses the (external) Z3 C API, which is in a private repo

Thanks again for your responses!

**Comment by NikolajBjorner** [comment]
If you are somehow losing axioms you have asserted it could be because they get removed during a pop.
context::pop_scope_core removes auxiliary clauses on the stack.
Using the external API for such extensions is generally preferred for several reasons:
1. The legacy core is really difficult to debug.
2.  I would like to sunset it in favor of a more extensible and cleaner core (it will still take several months before this is solid, though).
The user propagator abstraction is supported for both legacy and new core.

The big catch in your case is that I don't support user propagators for string sorts because there is no notion of when a string variable gets assigned a constant string. It could however track equalities without tracking assignments.

As you have your tentacles deep inside the solver already you do have access to the internal state of assertions. This state is accessed by the "copy" method and you can determine which parts should be relevant for you.

**Reply by mister-walter** [reply]
Nikolaj,

Thanks again for your reply! This is very helpful. The user propagator feature looks exciting, but I believe we do need information about string assignments (we'll investigate this more).

Just to be crystal clear, we're currently using `Z3_translate_solver` to copy a solver for use with the Lisp code (which uses the C API). This method results in calls to `context::pop_scope_core`, so it looks like this method is not going to work for us. It sounds like you're saying that we can use the legacy internal code to manually copy all of the relevant information over, which should work for us.

I also don't know how we can get our hands on the solver inside of e.g. our custom `theory`'s `new_eq_eh` method. We're currently using a hack (see [here](https://github.com/ankitku/z3/blob/d0e1f40d9fa2e0f6f87315f8e25479f415258b75/src/api/api_parsers.cpp#L174) and [here](https://github.com/ankitku/z3/blob/be833b34ee6686ea4e7dc2a3871612691856d0ba/src/tactic/portfolio/smt_strategic_solver.cpp#L169)) to get the solver that `Z3_eval_smtlib2_string` is using, but it's causing other problems.

Concretely, we have two questions:

1. Is it possible to get the current `solver` object from inside of a theory? (in our case the `solver` is not externally settable or available, because we're calling `Z3_eval_smtlib2_string`) Is there any particular reason that `smt::theory` doesn't provide a `get_solver` method?
2. Is it possible to create a copy of a `solver`/`context` in the middle of a `check_sat` that contains at least the externally added assertions + theory axioms? For our purposes, it doesn't matter if this new `solver`/`context` properly copies scopes, or if it just throws everything into the root scope.

I think you've already answered 2 with "use the legacy `copy` method".

**Reply by c-cube** [reply]
Sorry to hijack this, I'm curious, how does Z3 handle local clauses (stack scoped)? I've found that it's hard to remove clauses from 2-watch structures in minisat-style solvers. More generally, learning how z3 does this kind of low level manipulations would be super interesting

**Reply by NikolajBjorner** [reply]
1. No the "solver" object is an abstraction. It is not available inside the smt_context.
2. yes: since you have your callbacks inside the code you can do anything within one of your callbacks.

@c-cube: two-watch removal just means you traverse all literals in a removed clause and remove pointers to that clause in, if any, in the watch sets. With the invariant that only the two first literals in a clause are watched you actually only have to check the first two literals. The entire store about creating clauses within a scope has more to it. See the sat_solver.cpp for the reinit_stack logic. Since I also create fresh Boolean variables during search there is additional handling for ensuring variables are tracked. Normal SAT solvers don't have this requirement.

**Reply by c-cube** [reply]
I do get that, my question is more: if a watch set is a vector, you have to traverse all of it :-). Maybe in practice they are small though. Or do you use actual sets?

**Reply by NikolajBjorner** [reply]
just vectors that are walked. Overhead does not show up on profiler. 
Propagation walks vectors all the time many times over.

**Reply by c-cube** [reply]
thank you, that's useful information :)

**Reply by NikolajBjorner** [reply]
@c-cube: I didn't really represent everything: we have ways to tag clauses as removed:
```
                if (c.was_removed() || c.size() == 1 || c[1] != not_l) {
                    // Remark: this method may be invoked when the watch lists are not in a consistent state,
                    // and may contain dead/removed clauses, or clauses with removed literals.
                    // See: method propagate_unit at sat_simplifier.cpp
                    // So, we must check whether the clause was marked for deletion, or
                    // c[1] != not_l
                    *it2 = *it;
                    it2++;
                    break;
                }
```

and the simplifier re-constructs all watch lists in one sweep and therefore doesn't remove clauses.

**Reply by NikolajBjorner** [reply]
@mister-walter: maybe consider what it would take to expose enough callbacks for your functionality, similar to user_propagator callbacks. Then you don't ffi link z3 code with yours and you have interface boundaries better defined. This is more up-front overhead than the code you pointed to, but you can use user_propagator as boiler plate to replicate and you will not have to hack your way around what is an external representation of AST objects vs internal. Your LISP code would just see external.

**Reply by ankitku** [reply]
@NikolajBjorner thanks for suggesting user_propagator. It seems like a very clean way to integrate our decision procedure, and doesn't require ffi linking. I have some questions about the user_propagator class. 
I have uploaded a sample file here : https://github.com/ankitku/z3/blob/master/examples/ex.py
It solves the string equation xcyczvycya = yacwazvbu (x,y,z,u,v,w are variables). The function my_eq is not called with the ids assigned for the lhs and rhs. Or it may be that I do not have the correct ids? In which case, is it possible to get the correct ids?
Also, is it possible to extend user_propagator such that one can track integer variables as well?
Thanks

**Reply by NikolajBjorner** [reply]
user_propagator only works for bit-vectors and Boolean. I mention this above. The solver doesn't assign values to integers or strings during search. It only has to establish that there is a satisfying assignment and the temporary values to integers can change even without backtracking. You would have to either augment user_propagator or create your own version that takes care of the FFI portions you are using.

**Reply by NikolajBjorner** [reply]
To say it in a different way: the only theories where it suffices to rely on fixed_eh are bit-vectors and booleans. For integers you may be able to propagate based on watching equalities and disequalities but something in addition may be needed. Instead of exposing a general purpose API across theories, user_propagator is focused on the types where I can assume some value assignment.

**Reply by ankitku** [reply]
Thanks, this is helpful. I will look into augmenting user_propagator with int and string types.
